File: src/main/java/shane/leetcode/problems/easy/Q804UniqueMorseCodeWords.java
package shane.leetcode.problems.easy;

import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

public class Q804UniqueMorseCodeWords {

    @Test
    public void test() {
        assertThat(uniqueMorseRepresentations(new String[]{"gin", "zen", "gig", "msg"})).isEqualTo(2);
    }

    static final String[] arr = {".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."};

    public int uniqueMorseRepresentations(String[] words) {
        Set<String> set = new HashSet<>();
        for (String word : words) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < word.length(); i++) {
                sb.append(arr[word.charAt(i) - 'a']);
            }
            set.add(sb.toString());
        }

        return set.size();
    }

}

File: src/main/kotlin/shane/leetkode/problems/easy/Q804UniqueMorseCodeWords.kt
package shane.leetkode.problems.easy

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test

/**
 * Runtime: 190 ms, faster than 95.24% of Kotlin online submissions for Unique Morse Code Words.
Memory Usage: 35.2 MB, less than 71.43% of Kotlin online submissions for Unique Morse Code Words.
 */
class Q804UniqueMorseCodeWords {

    @Test
    fun test() {
        Assertions.assertThat(uniqueMorseRepresentations(arrayOf("gin", "zen", "gig", "msg"))).isEqualTo(2)
    }

    fun uniqueMorseRepresentations(words: Array<String>): Int {
        val morse = arrayOf(
            ".-",
            "-...",
            "-.-.",
            "-..",
            ".",
            "..-.",
            "--.",
            "....",
            "..",
            ".---",
            "-.-",
            ".-..",
            "--",
            "-.",
            "---",
            ".--.",
            "--.-",
            ".-.",
            "...",
            "-",
            "..-",
            "...-",
            ".--",
            "-..-",
            "-.--",
            "--.."
        )
        return words.map { s ->
            val sb = StringBuilder()
            for (c in s) {
                sb.append(morse[c - 'a'])
            }
            sb.toString()
        }.toCollection(hashSetOf()).size
    }

}
